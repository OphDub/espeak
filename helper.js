export const verbAndParse = async (verb, url, data) => {
  const initialFetch = await fetch(url, {
    body: JSON.stringify(data),
    cache: 'no-cache',
    headers: {
      'content-type': 'application/json'
    },
    method: `${verb}`
  });
  const userData = await initialFetch.json()
  console.log(userData);
  return userData;
};