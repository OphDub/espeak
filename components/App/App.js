import React from 'react';
import { StyleSheet, View } from 'react-native';
import { TabNavigator } from 'react-navigation';
import AwesomeAlert from 'react-native-awesome-alerts';
import firebase from 'react-native-firebase';

import Login from '../Login/Login';
import Settings from '../Settings/Settings';
import WordStackNav from '../WordStackNav/WordStackNav';
import { verbAndParse } from '../../helper';
import config from '../../fbconfig';

const routeConfig = {
  Decks: {
    screen: WordStackNav
  },
  Settings: {
    screen: Settings
  },
};

const navConfig = {
  tabBarOptions: {
    labelStyle: {
      fontSize: 22,
    },
    activeTintColor: '#3D3B3E',
    activeBackgroundColor: '#CFAAF1',
    inactiveTintColor: 'white',
    style: {
      backgroundColor: '#B47AEA',
    },
    tabStyle: {
      marginVertical: 10,
    },
  }
};

const RootNav = TabNavigator(routeConfig, navConfig);

firebase.initializeApp(config);

const auth = firebase.auth();

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      user: null,
      loading: false,
      showAlert: false,
      alertMsg: '',
      decks: [],
    };
  }

  componentDidMount = async () => {
    await this.getDecks();
  }

  showAlert = () => {
    this.setState({ showAlert: true });
  }

  hideAlert = () => {
    this.setState({ showAlert: false, loading: false });
  }

  handlePoints = () => {
    const { user } = this.state;

    user.points += 10;

    this.setState({ user });
  }

  handleHintPoints = () => {
    const { user } = this.state;

    user.points -= 5;

    this.setState({ user });
  }

  updateUser = async () => {
    try {
      const stack_id = this.updateUserStack();
      const { points, firebase_id } = this.state.user;
      const url = `https://espeak-be-opa.herokuapp.com/api/v1/users/${firebase_id}`;

      await verbAndParse('PATCH', url, { points, stack_id });
      this.setState({ user: {...this.state.user, stack_id }});
    } catch (error) {
      this.setState({
        showAlert: true,
        alertMsg: error.message
      });
    }
  }

  updateUserStack = () => {
    let { user } = this.state;
    let { stack_id } = user;
    const lastStack = this.state.decks[this.state.decks.length - 1];

    if (stack_id < lastStack.id) {
      stack_id++;
    }

    return stack_id;
  }

  handleLogin = async (email, password) => {
    this.setState({ loading: true });

    try {
      const firebaseInfo = await auth.signInAndRetrieveDataWithEmailAndPassword(email, password);
      await this.beLogin(firebaseInfo.user.uid);
    } catch (error) {
      this.setState({
        showAlert: true,
        alertMsg: error.message,
        loading: false
      });
    }
  }

  getDecks = async () => {
    const url = 'https://espeak-be.herokuapp.com/api/v1/stack';
    const decks = await verbAndParse('GET', url);
    await this.setState({ decks });
  }

  beLogin = async (userId) => {
    try {
      const initialFetch = await fetch(`https://espeak-be-opa.herokuapp.com/api/v1/users/${userId}`);
      const user = await initialFetch.json();
      
      this.setState({ user: user[0], loading: false });
    } catch (error) {
      this.setState({
        showAlert: true,
        alertMsg: 'Invalid user name or password.',
        loading: false
      });
    }
  }

  handleSignOut = () => {
    auth.signOut();
    this.setState({user: null});
  }

  handleRegistration = async (userInfo) => {
    const { email, userName, password } = userInfo;
    this.setState({ loading: true });

    try {
      const user = await auth.createUserWithEmailAndPassword(email, password);

      await this.beRegistration(user, userName);
    } catch (error) {
      this.setState({
        showAlert: true,
        alertMsg: error.message,
        loading: false
      });
    }
  }

  beRegistration = async (userInfo, userName) => {
    try {
      const { email, uid } = userInfo;
      const newUser = {
        name: userName,
        email: email,
        points: '0',
        stack_id: 1,
        firebase_id: uid
      };
      const url = 'https://espeak-be-opa.herokuapp.com/api/v1/users/';

      verbAndParse('POST', url, newUser);
      newUser.points = parseInt(newUser.points);
      this.setState({ user: newUser, loading: false });
    } catch (error) {
      this.setState({
        showAlert: true,
        alertMsg: error.message,
        loading: false
      });
    }
  }

  showCondition = () => {
    return this.state.user ?
      (<RootNav
        screenProps={{
          user: this.state.user,
          handleSignOut: this.handleSignOut,
          userPoints: this.state.user.points,
          handlePoints: this.handlePoints,
          handleHintPoints: this.handleHintPoints,
          updateUser: this.updateUser,
          decks: this.state.decks,
        }} 
      />) 
      :
      (<View style={styles.container}>
        <Login
          handleLogin={this.handleLogin}
          handleRegistration={this.handleRegistration}
          beRegistration={this.beRegistration}/>
        <AwesomeAlert
          show={this.state.showAlert}
          showProgress={false}
          title={'Uh oh! There is a problem!'}
          message={this.state.alertMsg}
          closeOnTouchOutside={true}
          closeOnHardwareBackPress={false}
          showCancelButton={false}
          showConfirmButton={true}
          confirmText="OK"
          confirmButtonColor="#3AAFb9"
          onConfirmPressed={() => {
            this.hideAlert();
          }}/>
        <AwesomeAlert 
          show={this.state.loading}
          showProgress={true}
        />
      </View>);
  }

  render() {
    return (
      this.showCondition()
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});
