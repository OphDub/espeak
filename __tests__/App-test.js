import React from 'react';
import { configure } from 'enzyme';
import App from '../components/App/App';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import firebase from 'react-native-firebase'
import { verbAndParse } from '../helper';
configure({ adapter: new Adapter() });

describe('App', () => {
  it('renders correctly', () => {
    const wrapper = renderer.create(<App />).toJSON();

    expect(wrapper).toMatchSnapshot();
  })

  it('should have default state', () => {
    const wrapper = renderer.create(<App />);
    const inst = wrapper.getInstance();

    expect(inst.state).toEqual({ "alertMsg": "", "decks": [], "loading": false, "showAlert": false, "user": null })
  })

  describe('App methods', () => {

    it('the showAlert function should toggle the state of showAlert', () => {
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();

      expect(inst.state.showAlert).toEqual(false);

      inst.showAlert();

      expect(inst.state.showAlert).toEqual(true);
    })

    it('the hideAlert function should toggle the state off', () => {
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();

      inst.setState({showAlert: true});

      expect(inst.state.showAlert).toEqual(true);

      inst.hideAlert();

      expect(inst.state.showAlert).toEqual(false);
    })

    it('the handlePoints function should add ten points to the user', () => {
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();
      const user = {
        username: 'pophus', 
        email: 'pophus@notpophanda.com', 
        points: 0, 
        stack_id: 1, 
        firebase_id: 12345
      };

      inst.setState({ user });

      expect(inst.state.user.points).toEqual(0);

      inst.handlePoints();

      expect(inst.state.user.points).toEqual(10);
    })

    it('the handleHintPoints function should deduct 5 points from the user', () => {
      const wrapper = renderer.create(<App />);
      const instance = wrapper.getInstance();
      const user = {
        username: 'ophanda', 
        email: 'ophanda@betterthanpophus.com', 
        points: 10, 
        stack_id: 1, 
        firebase_id: 12345
      };

      instance.setState({ user });

      expect(instance.state.user.points).toEqual(10);

      instance.handleHintPoints();

      expect(instance.state.user.points).toEqual(5);
    })
  
    it('the handleLogin function should set the loading state to true and call beLogin and update the state', async () => {
      const user = {
        username: 'pophus',
        email: 'pophus@notpophanda.com',
        points: 0,
        stack_id: 6,
        firebase_id: 12345
      };
      global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve([user])
      }))
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();

      await inst.handleLogin('pophus', 'password');

      expect(inst.state.loading).toEqual(false);
      expect(inst.state.user).toEqual({ "email": "pophus@notpophanda.com", "firebase_id": 12345, "points": 0, "stack_id": 6, "username": "pophus" })
    })

    it('beLogin call fetch and update the user in state', async () => {
      const user = {
        username: 'pophus',
        email: 'pophus@notpophanda.com',
        points: 0,
        stack_id: 1,
        firebase_id: 12345
      };
      global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () => Promise.resolve([user])
      }));
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();

      expect(inst.state.user).toEqual(null);
      await inst.beLogin(2);

      expect(inst.state.user).toEqual(user)
    })

    it('handleSignOut should set the user to null', () => {
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();
      const user = {
        username: 'pophus',
        email: 'pophus@notpophanda.com',
        points: 0,
        stack_id: 1,
        firebase_id: 12345
      };

      inst.setState({user});

      expect(inst.state.user).toBeDefined();

      inst.handleSignOut();

      expect(inst.state.user).toEqual(null);
    })

    it.only('handleRegistration should call beRegistration and update states user to the newly created user', async () => {
      global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        status: 200,
        json: () =>  Promise.resolve({user: 'pophus' } )
      }))
      const wrapper = renderer.create(<App />);
      const inst = wrapper.getInstance();

      await inst.handleRegistration({email: 'pophus@notpophanda.com', userName: 'pophus long', password: 'password', uid: 54321});

      expect(inst.state.user).toEqual({"email": "pophus@notpophanda.com", "firebase_id": 54321, "name": "pophus long", "points": 0, "stack_id": 6 })
    })
  })

})