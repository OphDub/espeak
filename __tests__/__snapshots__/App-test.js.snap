// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        eSpeak
      </Text>
      <Image
        source={
          Object {
            "uri": "https://i.imgur.com/ana1fGy.png",
          }
        }
        style={
          Object {
            "height": 100,
            "margin": 10,
            "width": 100,
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="email"
        style={
          Object {
            "backgroundColor": "#CFCCD0",
            "borderRadius": 8,
            "margin": 10,
            "padding": 10,
            "paddingHorizontal": 8,
            "paddingVertical": 10,
            "width": 200,
          }
        }
        value=""
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="password"
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#CFCCD0",
            "borderRadius": 8,
            "margin": 10,
            "padding": 10,
            "paddingHorizontal": 8,
            "paddingVertical": 10,
            "width": 200,
          }
        }
        value=""
      />
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3AAFB9",
            "borderRadius": 8,
            "display": "flex",
            "justifyContent": "center",
            "margin": 10,
            "opacity": 1,
            "paddingVertical": 10,
            "shadowColor": "#979797",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.7,
            "shadowRadius": 4,
            "width": 200,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
        >
          LOGIN
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": 25,
            "justifyContent": "center",
            "margin": 10,
            "opacity": 1,
            "width": 200,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          New? Register here
        </Text>
      </View>
    </View>
  </View>
</View>
`;
